src/controllers/queryRunner.ts

```diff
@@ -337,99 +351,88 @@ export default class QueryRunner {
+            if (process.platform === 'darwin') {
+                oldLang = process.env['LANG'];
+                process.env['LANG'] = 'en_US.UTF-8';
+            }
+            await this._vscodeWrapper.clipboardWriteText(copyString);
+            if (process.platform === 'darwin') {
+                process.env['LANG'] = oldLang;
             }
-            p.then(() => {
-                let oldLang: string;
-                if (process.platform === 'darwin') {
-                    oldLang = process.env['LANG'];
-                    process.env['LANG'] = 'en_US.UTF-8';
+        }
+
+        // create a mapping of the ranges to get promises
+        let tasks = selection.map((range, i) => {
+            return async () => {
+                const result = await this.getRows(range.fromRow, range.toRow - range.fromRow + 1, batchId, resultId);
+                if (this.shouldIncludeHeaders(includeHeaders)) {
+                    let columnHeaders = this.getColumnHeaders(batchId, resultId, range);
+                    if (columnHeaders !== undefined) {
+                        copyString += columnHeaders.join('\t') + os.EOL;
+                    }
                 }
-                this._vscodeWrapper.clipboardWriteText(copyString).then(() => {
-                    if (process.platform === 'darwin') {
-                        process.env['LANG'] = oldLang;
+
+                // Iterate over the rows to paste into the copy string
+                for (let rowIndex: number = 0; rowIndex < result.resultSubset.rows.length; rowIndex++) {
+                    let row = result.resultSubset.rows[rowIndex];
+                    let cellObjects = row.slice(range.fromCell, (range.toCell + 1));
+                    // Remove newlines if requested
+                    let cells = this.shouldRemoveNewLines()
+                        ? cellObjects.map(x => this.removeNewLines(x.displayValue))
+                        : cellObjects.map(x => x.displayValue);
+                    copyString += cells.join('\t');
+                    if (rowIndex < result.resultSubset.rows.length - 1) {
+                        copyString += os.EOL;
                     }
-                    resolve();
-                });
-            });
+                }
+            };
         });
+
+        let p = tasks[0]();
+        for (let i = 1; i < tasks.length; i++) {
+            p = p.then(tasks[i]);
+        }
+        await p;
+        let oldLang: string;
+        if (process.platform === 'darwin') {
+            oldLang = process.env['LANG'];
+            process.env['LANG'] = 'en_US.UTF-8';
+        }
+        await this._vscodeWrapper.clipboardWriteText(copyString);
+        if (process.platform === 'darwin') {
+            process.env['LANG'] = oldLang;
+        }
     }
 
 
-    public toggleSqlCmd(): Thenable<boolean> {
+    public async toggleSqlCmd(): Promise<boolean> {
         const queryExecuteOptions: QueryExecutionOptions = { options: new Map<string, any>() };
```

Comment:
nit - but it's weird to have this return a bool when it'll never be able to return a false as it's currently written